### DISCUSSION TITLE
How to add an array of timedeltas with cftime index?

### DISCUSSION THREAD

**Discussion by ahuang11** [pydata/xarray#4999] [Q&A]
This raises an error
```
import cftime
import numpy as np
import xarray as xr
import pandas as pd
da = xr.DataArray(
    data=xr.CFTimeIndex([cftime.DatetimeProlepticGregorian(1, 1, 1)])
)
td = pd.to_timedelta(['2H', '3H']).values

da + td
```

`UFuncTypeError: ufunc 'add' cannot use operands with types dtype('O') and dtype('<m8[ns]')`

Is the only solution to loop through?

**Comment by spencerkclark** [comment]
Oops I seemed to have missed this question.  To add timedeltas to cftime objects, you must use `datetime.timedelta` objects instead of `timedelta64[ns]` values (this is what cftime objects use for timedelta arithmetic internally).  E.g. the following works for your last line:
```python
da + td.to_pytimedelta()
```
However, there is currently no way to store `datetime.timedelta` objects directly within a DataArray (xarray will automatically cast them to `timedelta64[ns]` values).  The above example happens to work because we are adding a NumPy array of `datetime.timedelta` objects to a DataArray.  

This is not ideal, but I do not know what the best solution would be.  Similar to [what we have in `CFTimeIndex` itself](https://github.com/pydata/xarray/blob/e6168ed6b771b7377e89acb47cc7e5aa9f21a574/xarray/coding/cftimeindex.py#L553-L577), we could add logic to cast `timedelta64[ns]` values to `datetime.timedelta` objects when doing addition or subtraction of timedelta-like DataArrays from cftime DataArrays.  This would be helpful in a lot of cases, but that said, it would not work for timedelta objects greater than ~292 years due to the limits on the range of `timedelta64[ns]` values.

**Reply by spencerkclark** [reply]
> This would be helpful in a lot of cases, but that said, it would not work for timedelta objects greater than ~292 years due to the limits on the range of `timedelta64[ns]` values.

With new developments in pandas (see more discussion in #7493) we may no longer need to worry about the nanosecond-precision limitation, which would conveniently eliminate this issue.
